pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java

```diff
@@ -598,4 +648,139 @@ private static void initSqlCommands(Config sqlConfig) {
+            super(request);
+            String appEngineSourceIp = request.getHeader(preferredHeaderForIP);
+            remoteAddr = request.getRemoteAddr();
+            if (StringUtils.isNotBlank(appEngineSourceIp)) {
+                remoteAddr = appEngineSourceIp;
+            }
+        }
+
+        @Override
+        public String getRemoteAddr() {
+            return remoteAddr;
+        }
+    }
+
+    private static class AppEngineServerFactory implements EmbeddedServerFactory {
+        private final JettyServerFactory serverFactory;
+        private String preferredIPSourceHeader;
+        private ThreadPool threadPool;
+        private boolean httpOnly = true;
+
+        public AppEngineServerFactory(JettyServerFactory serverFactory, String preferredIPSourceHeader) {
+            this.serverFactory = serverFactory;
+            this.preferredIPSourceHeader = preferredIPSourceHeader;
+        }
+
+        public EmbeddedServer create(Routes routeMatcher, StaticFilesConfiguration staticFilesConfiguration, ExceptionMapper exceptionMapper, boolean hasMultipleHandler) {
+            MatcherFilter matcherFilter = new MatcherFilter(routeMatcher, staticFilesConfiguration, exceptionMapper, false, hasMultipleHandler);
+            matcherFilter.init((FilterConfig)null);
+            SessionHandler handler = new JettyCustomRemoteAddrHeaderHandler(matcherFilter, this.preferredIPSourceHeader);
+            handler.getSessionCookieConfig().setHttpOnly(this.httpOnly);
+            return (new EmbeddedJettyServer(this.serverFactory, handler)).withThreadPool(this.threadPool);
+        }
+
+        public AppEngineServerFactory withThreadPool(ThreadPool threadPool) {
+            this.threadPool = threadPool;
+            return this;
+        }
+
+        public AppEngineServerFactory withHttpOnly(boolean httpOnly) {
+            this.httpOnly = httpOnly;
+            return this;
+        }
+    }
+
+    /**
+     * Lifted from {@link spark.embeddedserver.jetty.JettyHandler}, with one
+     * change to use {@link SourceIPHeaderRequestWrapper} to set
+     * the source IP properly in the appengine environment
+     */
+    public static class JettyCustomRemoteAddrHeaderHandler extends SessionHandler {
+        private javax.servlet.Filter filter;
+        private String preferredIPSourceHeader;
+
+        public JettyCustomRemoteAddrHeaderHandler(Filter filter, String preferredIPSourceHeader) {
+            this.filter = filter;
+            this.preferredIPSourceHeader = preferredIPSourceHeader;
+        }
+
+        public void doHandle(String target, org.eclipse.jetty.server.Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+            HttpRequestWrapper wrapper = new SourceIPHeaderRequestWrapper(request, preferredIPSourceHeader);
+            this.filter.doFilter(wrapper, response, null);
```

Comment:
Sure about the null argument?

Reply:
I'm sure I copy-pasted it from sparkjava decompiled code, for what that's worth.